### PR TITLE
Refactor IP filter to work with appconfig and copilot 

### PR DESCRIPTION
**Pull request by lgarvey**
Initial refactoring to use appconfig
    
    Work done so far:
    
    Added a test appconfig agent
    Refactored env var logic to support app environment level variables which take precedence, .e.g. $STAGING_IPFILTER_ENABLED will override $IPFILTER_ENABLED in the staging environment. 

    Added IPFILTER_ENABLED, PROTECTED_PATHS and PUBLIC_PATHS config to allow the IP filter to be turned off, or to protect only certain paths, or to exclude certain paths.

    Temporarily removed the shared token and basic auth logic - it will be reintroduced in a later PR.

    Minor restructuring of the test cases to separate out tests that focus on ipfilter logic, proxy logic, and configuration logic.
    
    To do:
    Add versioning and a version endpoint
    Reintroduce basic auth and shared token logic (another PR most likely)
    Fix test_slow_upload test (currently skipped)
    Rewrite README
    I removed gevent because it has compilation issues.  Should it be reintroduced?

### CODE REVIEW

File: auth.py [hunk 1]
```diff
@@ -0,0 +1,9 @@
+
+def constant_time_is_equal(a, b):
```
Comment (gabelton) [inline auth.py:2]: Where does this get used? Is it worth adding a comment to explain what's happening?

File: config.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
+from collections import UserDict
+
+from urllib.parse import urljoin
+import urllib3
+import yaml
+
+
+class Environ(UserDict):
```
Comment (gabelton) [inline config.py:8]: could add type hinting to some of these methods? Some are obvious, but `get_ipfilter_config` and `get_appconfig_configuration` maybe would benefit

File: auth.py (hunk 1 above)
Comment (lgarvey) [reply auth.py:2]: Ah, I've left it there so it can be used when I reintroduce the basic auth logic.

File: config.py (hunk 2 above)
Comment (lgarvey) [reply config.py:8]: OK, I'll add some hints

File: config.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+        if not value:
+            return []
+
+        return [v.strip() for v in value.split(",")]
+
+    def bool(self, key, /, default=None, allow_environment_override=False):
+        value = self.get_value(
+            key, default=default, allow_environment_override=allow_environment_override
+        )
```
Comment (WillGibson) [inline config.py:52]: This is repeated several times so probably worth extracting to a private method...

```
self.get_value(
    key, default=default, allow_environment_override=allow_environment_override
)
```

File: main.py [hunk 4]
```diff
@@ -1,236 +1,124 @@
...
-@app.route('/', defaults={'u_path': ''}, methods=['GET', 'POST', 'PUT', 'PATCH', 'DELETE', 'OPTIONS', 'HEAD'])
-@app.route('/<path:u_path>', methods=['GET', 'POST', 'PUT', 'PATCH', 'DELETE', 'OPTIONS', 'HEAD'])
+@app.route(
+    "/",
+    defaults={"u_path": ""},
+    methods=["GET", "POST", "PUT", "PATCH", "DELETE", "OPTIONS", "HEAD"],
+)
+@app.route(
+    "/<path:u_path>",
+    methods=["GET", "POST", "PUT", "PATCH", "DELETE", "OPTIONS", "HEAD"],
```
Comment (WillGibson) [inline main.py:54]: Maybe extract the methods list to a variable if it's the same for both.

File: main.py [hunk 5]
```diff
@@ -1,236 +1,124 @@
...
-        (
-            hostname_ok[i] and
-            ip_ok[i] and
-            (not shared_secrets[i] or any(shared_secret_ok[i])) and
-            (not basic_auths[i] or any(basic_auths_ok[i]))
+        and (
+            not app.config["PUBLIC_PATHS"]
+            or not any(
+                path.startswith(request.path) for path in app.config["PUBLIC_PATHS"]
+            )
```
Comment (WillGibson) [inline main.py:100]: They are miles apart in the diff here, but we could refactor this out to a private method and just pass in the `app.config` key.

**Comment by lgarvey** [comment]
We've created a separate repo uktrade/ip-filter for the AWS Copilot version of the IP Filter.

**Pull request closed without merging**